File: scripts/test-longevity-scroll.mjs
import { chromium } from "playwright";
import assert from "node:assert/strict";

const URL = "http://localhost:8080/longevity";
const VIEWPORTS = [
  { width: 390, height: 844 },
  { width: 768, height: 1024 },
  { width: 974, height: 738 },
  { width: 1440, height: 900 },
];
const ORDER = ["1", "2", "4", "6"];

const state = async (page) => page.locator("[data-story-card]").evaluate((card) => ({
  chapter: card.dataset.chapter,
  phase: card.dataset.storyPhase,
  opacity: Number.parseFloat(getComputedStyle(card).opacity),
  transform: getComputedStyle(card).transform,
  count: document.querySelectorAll("[data-story-card]").length,
  scrollY,
}));

const openScene = async (page) => {
  await page.goto(URL, { waitUntil: "domcontentloaded" });
  await page.evaluate(() => window.scrollTo(0, document.querySelector("#timeline")?.offsetTop ?? 0));
  await page.waitForFunction(() => document.querySelector("[data-story-card]")?.getAttribute("data-story-phase") === "idle");
};

const wheelStep = async (page, delta = 120) => {
  const viewport = page.viewportSize();
  await page.mouse.move((viewport?.width ?? 800) / 2, (viewport?.height ?? 800) / 2);
  await page.mouse.wheel(0, delta);
};

const waitIdle = (page) => page.waitForFunction(() => document.querySelector("[data-story-card]")?.getAttribute("data-story-phase") === "idle");

const checkAnimation = async (page) => {
  const before = await state(page);
  const startedAt = performance.now();
  await wheelStep(page);
  await page.waitForTimeout(150);
  const leaving = await state(page);
  assert.equal(leaving.chapter, before.chapter, "Содержимое сменилось до завершения выхода");
  assert.equal(leaving.phase, "exit", "Через 150 ms должна идти фаза выхода");
  assert(leaving.opacity > 0.05 && leaving.opacity < 0.95, "Через 150 ms старая карточка должна оставаться видимой");

  await page.waitForFunction((chapter) => {
    const card = document.querySelector("[data-story-card]");
    return card?.getAttribute("data-chapter") !== chapter && Number.parseFloat(getComputedStyle(card).opacity) <= 0.05;
  }, before.chapter);
  const swapped = await state(page);
  assert.equal(swapped.count, 1, "В сцене должна быть ровно одна карточка");

  await page.waitForFunction(() => document.querySelector("[data-story-card]")?.getAttribute("data-story-phase") === "hold");
  const enteredAt = performance.now() - startedAt;
  assert(enteredAt >= 650 && enteredAt <= 900, `Вход завершился за ${Math.round(enteredAt)} ms вместо 650–900 ms`);
  const held = await state(page);
  await page.waitForTimeout(850);
  const heldLater = await state(page);
  assert.equal(heldLater.chapter, held.chapter, "Карточка сменилась во время паузы чтения");
  assert.equal(heldLater.opacity, held.opacity, "Opacity изменился во время паузы чтения");
  assert.equal(heldLater.transform, held.transform, "Положение изменилось во время паузы чтения");
  await waitIdle(page);
};

const checkWheelSequence = async (page) => {
  for (let position = 1; position < ORDER.length; position += 1) {
    await wheelStep(page);
    await waitIdle(page);
    assert.equal((await state(page)).chapter, ORDER[position]);
  }
  for (let position = ORDER.length - 2; position >= 0; position -= 1) {
    await wheelStep(page, -120);
    await waitIdle(page);
    assert.equal((await state(page)).chapter, ORDER[position]);
  }
};

const checkInertia = async (page) => {
  for (const delta of [14, 24, 36, 28, 20, 14, 12]) {
    await page.mouse.wheel(0, delta);
    await page.waitForTimeout(70);
  }
  await waitIdle(page);
  assert.equal((await state(page)).chapter, "2", "Инерционный импульс перескочил больше одной главы");
};

const checkOtherInputs = async (page) => {
  await page.keyboard.press("ArrowDown");
  await waitIdle(page);
  assert.equal((await state(page)).chapter, "4", "Клавиатура не переключила ровно одну главу");

  const slider = page.getByRole("slider", { name: "Перемотка таймлайна" });
  await slider.focus();
  await page.keyboard.press("End");
  await waitIdle(page);
  assert.equal((await state(page)).chapter, "6", "Ползунок не использовал общий переход");

  await page.evaluate(() => {
    const start = new TouchEvent("touchstart", { bubbles: true, touches: [{ clientY: 500 }] });
    const move = new TouchEvent("touchmove", { bubbles: true, cancelable: true, touches: [{ clientY: 560 }] });
    window.dispatchEvent(start);
    window.dispatchEvent(move);
    window.dispatchEvent(new TouchEvent("touchend", { bubbles: true }));
  });
  await waitIdle(page);
  assert.equal((await state(page)).chapter, "4", "Свайп не переключил ровно одну главу");
};

const checkBoundaries = async (page) => {
  const topBefore = await page.evaluate(() => scrollY);
  await wheelStep(page, -120);
  await page.waitForTimeout(250);
  const topAfter = await page.evaluate(() => scrollY);
  assert(topAfter < topBefore, "На первой главе прокрутка вверх должна уходить странице");

  await page.evaluate(() => window.scrollTo(0, document.querySelector("#timeline")?.offsetTop ?? 0));
  for (let index = 1; index < ORDER.length; index += 1) {
    await wheelStep(page);
    await waitIdle(page);
  }
  const bottomBefore = await page.evaluate(() => scrollY);
  await wheelStep(page);
  await page.waitForTimeout(250);
  const bottomAfter = await page.evaluate(() => scrollY);
  assert(bottomAfter > bottomBefore, "На последней главе прокрутка вниз должна уходить странице");
};

const checkViewport = async (browser, viewport, comprehensive) => {
  const context = await browser.newContext({ viewport });
  const page = await context.newPage();
  const errors = [];
  page.on("pageerror", (error) => errors.push(error.message));
  await openScene(page);

  const y = await page.evaluate(() => scrollY);
  await checkAnimation(page);
  assert(Math.abs((await page.evaluate(() => scrollY)) - y) <= 2, "Сцена прокрутила страницу во время внутреннего шага");
  const cardBox = await page.locator("[data-story-card]").boundingBox();
  const sceneBox = await page.locator("[data-timeline-scene]").boundingBox();
  assert(cardBox && sceneBox, "Не удалось измерить сцену");
  assert(cardBox.x >= sceneBox.x - 1 && cardBox.x + cardBox.width <= sceneBox.x + sceneBox.width + 1, "Карточка вышла за горизонтальные границы сцены");
  assert(cardBox.y >= sceneBox.y - 1 && cardBox.y + cardBox.height <= sceneBox.y + sceneBox.height + 1, "Карточка вышла за вертикальные границы сцены");
  assert.equal(await page.evaluate(() => document.documentElement.scrollWidth - document.documentElement.clientWidth), 0, "Появился горизонтальный скролл");
  assert.deepEqual(errors, [], `Ошибки страницы: ${errors.join("; ")}`);

  if (comprehensive) {
    await openScene(page);
    await checkWheelSequence(page);
    await openScene(page);
    await checkInertia(page);
    await checkOtherInputs(page);
    await openScene(page);
    await checkBoundaries(page);
  }
  await context.close();
};

const browser = await chromium.launch({ headless: true });
try {
  for (const viewport of VIEWPORTS) {
    await checkViewport(browser, viewport, viewport.width === 974);
    console.log(`PASS ${viewport.width}×${viewport.height}`);
  }
  console.log("PASS: все критерии скролл-сцены выполнены");
} finally {
  await browser.close();
}

File: scripts/test_longevity_scroll.py
"""Комплексная проверка скролл-сцены /longevity.

Запуск: python3 scripts/test_longevity_scroll.py
Сцена привязана к позиции прокрутки: главы 1 -> 2 -> 4 -> 6 не могут быть
пропущены ни на какой скорости, прокрутка страницы не перехватывается.
"""

import asyncio
import sys

from playwright.async_api import async_playwright

URL = "http://localhost:8080/longevity"
SIZES = [(390, 844), (768, 1024), (974, 738), (1440, 900)]
EXPECTED = ["1", "2", "4", "6"]
SPEEDS = [(300, 100), (200, 400), (120, 1200)]

STATE = """() => {
  const card = document.querySelector('[data-story-card]');
  const cards = document.querySelectorAll('[data-story-card]').length;
  const style = card ? getComputedStyle(card) : null;
  return {
    y: window.scrollY,
    cards,
    chapter: card ? card.getAttribute('data-chapter') : null,
    phase: card ? card.getAttribute('data-story-phase') : null,
    opacity: style ? parseFloat(style.opacity) : null,
    transform: style ? style.transform : null,
    label: document.querySelector('[data-timeline-scene] p')?.textContent ?? '',
    overflowX: document.documentElement.scrollWidth - document.documentElement.clientWidth,
  };
}"""


def dedupe(values):
    out = []
    for value in values:
        if not out or out[-1] != value:
            out.append(value)
    return out


async def scroll_pass(page, delta, gap, steps, failures, label):
    states = []
    blocked = 0
    previous_y = await page.evaluate("window.scrollY")
    for _ in range(steps):
        await page.mouse.wheel(0, delta)
        await page.wait_for_timeout(gap)
        state = await page.evaluate(STATE)
        states.append(state)
        at_bottom = await page.evaluate("window.scrollY + window.innerHeight >= document.documentElement.scrollHeight - 2")
        if delta > 0 and state["y"] == previous_y and state["y"] > 0 and not at_bottom:
            blocked += 1
        previous_y = state["y"]
        if state["cards"] != 1:
            failures.append(f"{label}: карточек в DOM {state['cards']}")
    sequence = dedupe([state["chapter"] for state in states if state["chapter"]])
    return states, sequence, blocked


async def check_size(page, width, height, errors, failures):
    await page.set_viewport_size({"width": width, "height": height})
    tag = f"{width}x{height}"
    await page.goto(URL, wait_until="domcontentloaded")
    await page.wait_for_timeout(1200)

    for delta, gap in SPEEDS:
        await page.evaluate("window.scrollTo(0, 0)")
        await page.wait_for_timeout(600)
        steps = max(int((6.0 * height) / delta) + 6, 12)
        label = f"{tag} вниз d={delta}/{gap}ms"
        states, sequence, blocked = await scroll_pass(page, delta, gap, steps, failures, label)
        if sequence != EXPECTED:
            failures.append(f"{label}: главы {sequence}, ожидалось {EXPECTED}")
        if blocked:
            failures.append(f"{label}: прокрутка заблокирована {blocked} раз")
        hold_by_chapter = {}
        for state in states:
            if state["phase"] == "hold" and state["chapter"]:
                hold_by_chapter[state["chapter"]] = hold_by_chapter.get(state["chapter"], 0) + 1
                if state["opacity"] is not None and state["opacity"] < 0.999:
                    failures.append(f"{label}: во время hold opacity {state['opacity']}")
        for chapter in EXPECTED:
            if chapter not in hold_by_chapter:
                failures.append(f"{label}: у главы {chapter} нет фазы удержания")
        for state in states:
            if state["overflowX"] != 0:
                failures.append(f"{label}: горизонтальный overflow {state['overflowX']}")
                break
        for state in states:
            if state["chapter"] and state["label"] and state["label"].strip():
                expected_range = {"1": "0", "2": "35", "4": "40", "6": "80"}[state["chapter"]]
                if width >= 640 and not state["label"].strip().startswith(expected_range):
                    failures.append(f"{label}: подпись «{state['label']}» не совпадает с главой {state['chapter']}")
                    break

    label = f"{tag} вверх"
    _, sequence, _ = await scroll_pass(page, -200, 400, max(int((6.0 * height) / 200) + 6, 12), failures, label)
    if sequence != list(reversed(EXPECTED)):
        failures.append(f"{label}: главы {sequence}, ожидалось {list(reversed(EXPECTED))}")

    if errors:
        failures.append(f"{tag}: ошибки страницы {errors[:3]}")


async def main():
    failures = []
    async with async_playwright() as playwright:
        browser = await playwright.chromium.launch(headless=True)
        context = await browser.new_context(viewport={"width": 1440, "height": 900})
        page = await context.new_page()
        errors = []
        page.on("pageerror", lambda error: errors.append(str(error)))
        for width, height in SIZES:
            errors.clear()
            await check_size(page, width, height, errors, failures)
            print(f"проверено {width}x{height}")
        await browser.close()

    if failures:
        print("\nПРОВАЛЫ:")
        for failure in failures:
            print(" -", failure)
        sys.exit(1)
    print("\nвсе проверки пройдены")


asyncio.run(main())
